src/rng/src/vhu_rng.rs

```diff
@@ -0,0 +1,315 @@
+        max_bytes: usize,
+    ) -> std::result::Result<Self, std::io::Error> {
+        Ok(VuRngBackend {
+            mem: None,
+            event_idx: false,
+            random_file,
+            timer: VuRngTimerConfig::new(period_ms, max_bytes),
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| VuRngError::EventFdError)?,
+        })
+    }
+
+    pub fn process_queue(&mut self, vring: &VringRwLock) -> Result<bool> {
+        // Copy the DescriptorChains to a vector.  Processing DescriptorChains
+        // one by one directly from the Vring would lead to a deadlock because
+        // the Vring's RwLock is taken once to access the DescriptorChains and
+        // once more when vring.add_used() is called.
+        let mut desc_chains: Vec<_> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter()
+            .map_err(|_| VuRngError::DescriptorNotFound)?
+            .collect();
+
+        for desc_chain in desc_chains.iter_mut() {
+            let descriptor = match desc_chain.next() {
+                Some(descriptor) => descriptor,
+                None => continue,
+            };
+
+            let mut to_read = descriptor.len() as usize;
+            let mut timer = &mut self.timer;
+            let mut random_file = self
+                .random_file
+                .lock()
+                .map_err(|_| VuRngError::UnexpectedRngSourceAccessError)?;
+
+            if !descriptor.is_write_only() {
+                return Err(VuRngError::UnexpectedReadDescriptor);
+            }
+
+            // Get the current time
+            let now = Instant::now();
+
+            // Check how much time has passed since we started the last period.
+            match now.checked_duration_since(timer.period_start) {
+                Some(duration) => {
+                    let elapsed = duration.as_millis();
+
+                    if elapsed >= timer.period_ms {
+                        // More time has passed than a full period, reset time
+                        // and quota.
+                        timer.period_start = now;
+                        timer.quota_remaining = timer.max_bytes;
+                    } else {
+                        // If we are out of bytes for the current period.  Block until
+                        // the start of the next period.
+                        if timer.quota_remaining == 0 {
+                            let to_sleep = timer.period_ms - elapsed;
+
+                            sleep(Duration::from_millis(to_sleep as u64));
+                            timer.period_start = Instant::now();
+                            timer.quota_remaining = timer.max_bytes;
+                        }
+                    }
+                }
+                None => return Err(VuRngError::UnexpectedTimerValue),
+            };
+
+            let mem = match &self.mem {
+                Some(m) => m.memory(),
+                None => return Err(VuRngError::NoMemoryConfigured),
```

Comment:
This error case is relatively straight forward to test as you just need to initialize `VuRngBackend` and call `process_queue` before calling the function that initializes the memory. Can you add a test for it? For the test to be easy to write, you'll need to derive `PartialEq` and `Debug` on the `Error`.

Reply:
In order to test this condition a Vring populated with DescriptorChains that points to valid Descriptors would need to be provided, and that is so tricky to do that the vhost-user-backend crate doesn't have that kind of test in its test suite.  

How should I proceed?

Reply:
Can you take a look at how we are testing virtio-queue in vm-virtio? There is an example here: https://github.com/rust-vmm/vm-virtio/blob/d62f2246568d4f544e848b23c025b268effac5ca/crates/virtio-queue/src/lib.rs#L1144
Do you think any of that work can be reused to create a test descriptor? We could also open an issue in vhost-user-backend if some kind of support is needed there.

If this turns out to be more complicated than initially expected, we can open an issue and address it in a future PR. There is though a higher risk of introducing bugs as we move forward, so we should make sure we have the appropriate primitives exported in vhost-user-backend so that we can test devices created on top of it.

Reply:
I'm pretty sure the virtio-queue crate has all the building blocks we need - let's see what we can do with that...